### ISSUE TITLE
Arango cluster in docker

### ISSUE THREAD

**Issue by virus2016** [arangodb/arangodb-docker#21]
Hi guys,

Bit of help please. I'm wanting to build an arrango cluster in docker. 10 data nodes and 3 coordinators. Firstly, is this possible? Or I'm I doing something wrong... (Must be the elasticsearch within me)

Thanks in advance

**Comment by fceller** [comment]
@virus2016 yes, you can build a cluster with different numbers of data nodes and coordinator nodes. We provide different scripts to deploy ArangoDB in the cloud. For details see https://github.com/ArangoDB/deployment

There are scripts to deploy ArangoDB on DigitalOcean, AWS, Azure and Google GCE.

If you want to test a cluster on your local machine, you can follow the instruction given in https://github.com/m0ppers/mesos-cluster

This will setup a cluster locally using mini-mesos and docker.

Is that, what you had in mind?

**Comment by virus2016** [comment]
Thanks Frank!

Is there a way of self discovery like elasticsearch? And do the node only replicate data i.e. each node has the same data on it?

Many thanks,
Calum MacRae

-------- Original message --------

**Comment by fceller** [comment]
The replication of data depends on your configuration. Normally you would set up a cluster, in which the collections are shared. I. e. each node only contains part of the data. In the standard setup, every shared is stored on two servers for resilience. You could also configure the cluster in such a way, that there is only one shard in which case each server will have the same data on it. What is your use-case? Do you scale because you have so much data, that it does not fit on a single server? Or do you need to scale because you need more CPU power for queries?

Currently, we are using Mesos for discovery. Mesos handles the resource management for us and sets up service locators to automatically find the service endpoints. We are currently working hard to simplify the process even further. In the near future, we will also support Kubernetes. In which environment (cloud or otherwise) are you planning to us ArangoDB?

Best
  Frank

**Comment by virus2016** [comment]
Hey Frank,

Thanks for spending time messaging me back.

We are currently using Rancher UI which is a docker orchestration tool (Awesome BTW). I am looking to scale for the CPU side. The data isn’t that big.

So let’s make it simple. I am wanting to create 10 “data” and 3 “client” containers. How would I go about doing that purely from the command line i.e. docker run….

Thanks again for your help on this.
Cheers,

**Comment by fceller** [comment]
I'm not familiar with Rancher UI - to be honest. How does fail-over and restart of failed containers? Is there anything elaborated like Apache Mesos Resource scheduling? Or does it "just" start a number of containers? If so, how are IP addresses of these containers published?

In General: you can start ArangoDB as a bunch of docker containers manually. A good starting point is

https://github.com/arangodb/deployment/blob/master/Docker/ArangoDBClusterWithDocker.sh

this script will deploy ArangoDB as docker containers.

Best
  Frank

**Comment by virus2016** [comment]
Hey Frank,

Now that 3.0 is out, is there away to create a cluster from the docker image?

Cheers,
Cal

**Comment by dothebart** [comment]
Hi @virus2016, 
We have added a tiny howto setup a temporary Mesosphere docker cluster with Mesosphere:

https://github.com/arangodb/arangodb/blob/devel/README_maintainers.md#arangodb-on-mesos 

(please note that this is a temporary instance, which will vanish if you terminate the controlling process, but it should serve well how to get started)

**Comment by virus2016** [comment]
Hi @dothebart,

I've used the mesos setup but I'm looking to try and simplify this deployment. Our customers will not run Mesos... 

I'm I right in thinking that each role should be a separate container in docker? DBServer, agency and coordinator? If so, can't we just do docker run x3 and what commands would they be?

Cheers,

**Comment by dothebart** [comment]
Hi @virus2016, meanwhile we used the time to finish the documentation for the cluster setup:
https://docs.arangodb.com/3.0/Manual/Deployment/index.html
Please note that i.e. automatic failover is implemented by mesosphere, which you will have to do by hand if you roll your own.

**Comment by lavvy** [comment]
Hey can't this be deployed with docker swarm?

**Comment by virus2016** [comment]
Another quick question, could someone create a docker-compose file for deployment of a cluster? That would help! 

Cheers,
Cal

**Comment by virus2016** [comment]
@fceller Can you help?

Cheers,

**Comment by dothebart** [comment]
we currently have https://github.com/arangodb/arangodb-mesos-docker - maybe that would be a good starting point for you? Since we don't exactly know whats expected there you would need to help us out here. If you want more interactivity to solve this, maybe join our slack channel so we can discuss this further how you could solve this?

**Comment by fceller** [comment]
Hi @virus2016 if you stay with us a little time longer, I think we can just do these things very easily. We are currently streamlining the boot process considerably so that starting in a non-mesos environment becomes almost trivial.

**Comment by virus2016** [comment]
Thanks guys, this will really help us! In the meantime, here is my docker-compose. Can you have a look over it and see if I am doing the compose right?

`version: "2"


services:

    agency: 
        image: arangodb/arangodb
        environment:
            - ARANGO_NO_AUTH=1
        command: arangod --server.endpoint tcp://0.0.0.0:5001 --server.authentication false --agency.activate true --agency.size 1 --agency.supervision true --database.directory /var/lib/arangodb3/agency1

    coordinator:
        image: arangodb/arangodb
        environment:
            - ARANGO_NO_AUTH=1
        command: arangod --server.authentication=false --server.endpoint tcp://0.0.0.0:8529 --cluster.my-address tcp://coordinator:8529 --cluster.my-local-info coord1  --cluster.my-role COORDINATOR --cluster.agency-endpoint tcp://agency:5001   --database.directory /var/lib/arangodb3/coordinator
        ports: ['8000:8529']
        depends_on:
            - agency
    coordinator2:
        image: arangodb/arangodb
        environment:
            - ARANGO_NO_AUTH=1
        command: arangod --server.authentication=false --server.endpoint tcp://0.0.0.0:8529 --cluster.my-address tcp://coordinator2:8529 --cluster.my-local-info coord2  --cluster.my-role COORDINATOR --cluster.agency-endpoint tcp://agency:5001   --database.directory /var/lib/arangodb3/coordinator
        ports: ['8001:8529']
        depends_on:
            - agency

    db1:
        image: arangodb/arangodb
        environment:
            - ARANGO_NO_AUTH=1
        command: arangod --server.authentication=false --server.endpoint tcp://0.0.0.0:8529 --cluster.my-address tcp://db1:8529 --cluster.my-local-info db1 --cluster.my-role PRIMARY --cluster.agency-endpoint tcp://agency:5001  --database.directory /var/lib/arangodb3/primary1
        depends_on:
            - agency
    db2:
        image: arangodb/arangodb
        environment:
            - ARANGO_NO_AUTH=1
        command: arangod --server.authentication=false --server.endpoint tcp://0.0.0.0:8529 --cluster.my-address tcp://db2:8529 --cluster.my-local-info db2 --cluster.my-role PRIMARY --cluster.agency-endpoint tcp://agency:5001  --database.directory /var/lib/arangodb3/primary1
        depends_on:
            - agency

    db3:
        image: arangodb/arangodb
        environment:
            - ARANGO_NO_AUTH=1
        command: arangod --server.authentication=false --server.endpoint tcp://0.0.0.0:8529 --cluster.my-address tcp://db3:8529 --cluster.my-local-info db3 --cluster.my-role PRIMARY --cluster.agency-endpoint tcp://agency:5001  --database.directory /var/lib/arangodb3/primary1
        depends_on:
            - agency`

**Comment by neunhoef** [comment]
I am starting to look now. Initial observation: I think your problems might be related to Docker/port issues. I do not know docker-compose too well, but as far as I see, your agency entry does not make sure that the agency is visible on port 5001 outside its docker container. I will experiment to find out.

**Comment by neunhoef** [comment]
The agency should work provided you forward port 5001 to outside the docker container, either by adding `ports: ['5001:5001']` or specifying the equivalent for docker-compose of `--net=host`.

**Comment by neunhoef** [comment]
I think I withdraw all my comments. Here is the file `docker-compose.yml` which I have used and which works for me: https://gist.github.com/neunhoef/1620c6c50e84a12be2b476bed419c644
This is essentially your file, I was doing `docker pull arangodb/arangodb` first to make sure I use version 3.1.3. It immediately worked for me. What problems did you have?
I use the following versions on Ubuntu 16.04: 
```
neunhoef@zen:~/dockercompose$ docker-compose --version
docker-compose version 1.9.0, build 2585387
neunhoef@zen:~/dockercompose$ docker --version
Docker version 1.12.3, build 6b644ec
```

**Comment by virus2016** [comment]
@neunhoef sorry yes it does work but I wasn't clear. I guess I was asking @fceller if that is the best way today to create the cluster. If it is I think you should commit that to the documentation as I'm sure others will have similar issues with this.

**Comment by neunhoef** [comment]
I think this setup is all right. It is good that only the coordinator ports are exposed on the host's main interface. I like that the others find themselves by using docker-composes' name lookup. Obviously, as docker-compose, it only works on a single machine, but for testing this is sensible. I will suggest to put this as an additional section in the Deployment chapter of the manual, because I think this can be useful for other people.

**Comment by virus2016** [comment]
@neunhoef I am having some issues with the cluster in docker. Maybe you can point me in the right direction. 

I am running the following:

    version: "2"
    services:

        agency: 
            image: arangodb/arangodb
            environment:
                - ARANGO_NO_AUTH=1
            command: arangod --server.endpoint tcp://0.0.0.0:8529 --agency.my-address tcp://agency:8529 --server.authentication false --agency.size 3 --agency.activate true --agency.supervision true  --database.directory /var/lib/arangodb3/agency1
            depends_on:
                - agency3
        agency2: 
            image: arangodb/arangodb
            environment:
                - ARANGO_NO_AUTH=1
            command: arangod --server.endpoint tcp://0.0.0.0:8529 --agency.my-address tcp://agency2:8529 --server.authentication false --agency.size 3 --agency.activate true --agency.supervision true --database.directory /var/lib/arangodb3/agency2
            depends_on:
                - agency3

        agency3: 
            image: arangodb/arangodb
            environment:
                - ARANGO_NO_AUTH=1
            command: arangod --server.endpoint tcp://0.0.0.0:8529 --agency.my-address tcp://agency3:8529 --server.authentication false --agency.size 3 --agency.endpoint tcp://agency:8529 --agency.endpoint tcp://agency2:8529 --agency.endpoint tcp://agency3:8529 --agency.activate true --agency.supervision true --database.directory /var/lib/arangodb3/agency3



        coordinator:
            image: arangodb/arangodb
            environment:
                - ARANGO_NO_AUTH=1
            command: arangod --server.authentication=false --server.endpoint tcp://0.0.0.0:8529 --cluster.my-address tcp://coordinator:8529 --cluster.my-local-info coord1  --cluster.my-role COORDINATOR --cluster.agency-endpoint tcp://agency:8529 --cluster.agency-endpoint tcp://agency2:8529 --cluster.agency-endpoint tcp://agency3:8529  --database.directory /var/lib/arangodb3/coordinator
            ports: ['8000:8529']
            depends_on:
                - agency
                - agency2
                - agency3

        coordinator1:
            image: arangodb/arangodb
            environment:
                - ARANGO_NO_AUTH=1
            command: arangod --server.authentication=false --server.endpoint tcp://0.0.0.0:8529 --cluster.my-address tcp://coordinator1:8529 --cluster.my-local-info coord2  --cluster.my-role COORDINATOR --cluster.agency-endpoint tcp://agency:8529 --cluster.agency-endpoint tcp://agency2:8529 --cluster.agency-endpoint tcp://agency3:8529  --database.directory /var/lib/arangodb3/coordinator1
            ports: ['8001:8529']
            depends_on:
                - agency
                - agency2
                - agency3

        coordinator2:
            image: arangodb/arangodb
            environment:
                - ARANGO_NO_AUTH=1
            command: arangod --server.authentication=false --server.endpoint tcp://0.0.0.0:8529 --cluster.my-address tcp://coordinator2:8529 --cluster.my-local-info coord3  --cluster.my-role COORDINATOR --cluster.agency-endpoint tcp://agency:8529 --cluster.agency-endpoint tcp://agency2:8529 --cluster.agency-endpoint tcp://agency3:8529  --database.directory /var/lib/arangodb3/coordinator2
            ports: ['8002:8529']
            depends_on:
                - agency
                - agency2
                - agency3

        db1:
            image: arangodb/arangodb
            environment:
                - ARANGO_NO_AUTH=1
            command: arangod --server.authentication=false --server.endpoint tcp://0.0.0.0:8529 --cluster.my-address tcp://db1:8529 --cluster.my-local-info db1 --cluster.my-role PRIMARY --cluster.agency-endpoint tcp://agency:8529 --cluster.agency-endpoint tcp://agency2:8529 --cluster.agency-endpoint tcp://agency3:8529 --database.directory /var/lib/arangodb3/primary1
            depends_on:
                - agency
                - agency2
                - agency3
                - coordinator
                - coordinator2
                - coordinator1

        db2:
            image: arangodb/arangodb
            environment:
                - ARANGO_NO_AUTH=1
            command: arangod --server.authentication=false --server.endpoint tcp://0.0.0.0:8529 --cluster.my-address tcp://db2:8529 --cluster.my-local-info db2 --cluster.my-role PRIMARY --cluster.agency-endpoint tcp://agency:8529 --cluster.agency-endpoint tcp://agency2:8529 --cluster.agency-endpoint tcp://agency3:8529  --database.directory /var/lib/arangodb3/primary2
            depends_on:
                - agency
                - agency2
                - agency3
                - coordinator
                - coordinator2
                - coordinator1


It all starts up and works fine. I add a DB and it sometimes crashes due to a replication err. Now if I am lucky to create the DB, I then create a collection. 3 shards 2 replications. I then add some data (250,000 docs) then I get this err:

←[36;1mdb1_1           |←[0m 2016-12-08T00:47:35Z [1] INFO {replication} connected to master at tcp://db2:8529, id 122177945464438, version 3.1, last log tick 2102976
←[31mcoordinator_1   |←[0m 2016-12-08T00:58:55Z [1] WARNING {agencycomm} Got a redirect 307 from agency endpoint: http+tcp://agency2:8529 location: http://agency:8529/_api/agency/read new forced endpoint: http+tcp://agency:8529/_api/agency/read
←[31mcoordinator_1   |←[0m 2016-12-08T00:58:55Z [1] WARNING {agencycomm} Retrying agency communication at http+tcp://agency:8529 tries: 2
←[36;1mdb1_1           |←[0m 2016-12-08T00:58:55Z [1] WARNING {agencycomm} Waiting on leader election http+tcp://agency2:8529 tries: 0
←[34mcoordinator2_1  |←[0m 2016-12-08T00:58:55Z [1] WARNING {agencycomm} Got a redirect 307 from agency endpoint: http+tcp://agency2:8529 location: http://agency:8529/_api/agency/read new forced endpoint: http+tcp://agency:8529/_api/agency/read
←[34mcoordinator2_1  |←[0m 2016-12-08T00:58:55Z [1] WARNING {agencycomm} Retrying agency communication at http+tcp://agency:8529 tries: 2
←[33;1mdb2_1           |←[0m 2016-12-08T00:58:55Z [1] WARNING {agencycomm} Got a redirect 307 from agency endpoint: http+tcp://agency2:8529 location: http://agency:8529/_api/agency/read new forced endpoint: http+tcp://agency:8529/_api/agency/read
←[33;1mdb2_1           |←[0m 2016-12-08T00:58:55Z [1] WARNING {agencycomm} Retrying agency communication at http+tcp://agency:8529 tries: 2
←[35mcoordinator1_1  |←[0m 2016-12-08T00:58:55Z [1] WARNING {agencycomm} Got a redirect 307 from agency endpoint: http+tcp://agency2:8529 location: http://agency:8529/_api/agency/write new forced endpoint: http+tcp://agency:8529/_api/agency/write
←[35mcoordinator1_1  |←[0m 2016-12-08T00:58:55Z [1] WARNING {agencycomm} Retrying agency communication at http+tcp://agency:8529 tries: 2
←[36;1mdb1_1           |←[0m 2016-12-08T00:58:56Z [1] WARNING {agencycomm} Got a redirect 307 from agency endpoint: http+tcp://agency2:8529 location: http://agency:8529/_api/agency/write new forced endpoint: http+tcp://agency:8529/_api/agency/write
←[36;1mdb1_1           |←[0m 2016-12-08T00:58:56Z [1] WARNING {agencycomm} Retrying agency communication at http+tcp://agency:8529 tries: 2
←[34mcoordinator2_1  |←[0m 2016-12-08T01:01:34Z [1] WARNING {agencycomm} Got a redirect 307 from agency endpoint: http+tcp://agency2:8529 location: http://agency:8529/_api/agency/write new forced endpoint: http+tcp://agency:8529/_api/agency/write
←[34mcoordinator2_1  |←[0m 2016-12-08T01:01:34Z [1] WARNING {agencycomm} Retrying agency communication at http+tcp://agency:8529 tries: 2
←[36;1mdb1_1           |←[0m 2016-12-08T01:02:53Z [1] INFO {replication} connected to master at tcp://db2:8529, id 122177945464438, version 3.1, last log tick 3188196
←[33;1mdb2_1           |←[0m 2016-12-08T01:02:53Z [1] INFO {replication} connected to master at tcp://db1:8529, id 180672683468752, version 3.1, last log tick 2115228
←[36;1mdb1_1           |←[0m 2016-12-08T01:02:54Z [1] INFO {replication} connected to master at tcp://db2:8529, id 122177945464438, version 3.1, last log tick 3188209
←[33;1mdb2_1           |←[0m 2016-12-08T01:02:56Z [1] INFO {replication} connected to master at tcp://db1:8529, id 180672683468752, version 3.1, last log tick 2115603
←[36;1mdb1_1           |←[0m 2016-12-08T01:02:57Z [1] INFO {replication} connected to master at tcp://db2:8529, id 122177945464438, version 3.1, last log tick 3188598
←[33;1mdb2_1           |←[0m 2016-12-08T01:03:09Z [1] ERROR {replication} insertLocal: dropping follower DBServer001 for shard s100078
←[36;1mdb1_1           |←[0m 2016-12-08T01:03:09Z [1] ERROR {replication} insertLocal: dropping follower DBServer002 for shard s100075
←[33;1mdb2_1           |←[0m 2016-12-08T01:03:09Z [1] ERROR {replication} insertLocal: dropping follower DBServer001 for shard s100076
←[33;1mdb2_1           |←[0m 2016-12-08T01:03:09Z [1] ERROR {replication} insertLocal: dropping follower DBServer001 for shard s100074
←[36;1mdb1_1           |←[0m 2016-12-08T01:03:09Z [1] ERROR {replication} insertLocal: dropping follower DBServer002 for shard s100077
←[31mcoordinator_1   |←[0m 2016-12-08T01:03:09Z [1] WARNING {agencycomm} Retrying agency communication at http+tcp://agency2:8529 tries: 2
←[31mcoordinator_1   |←[0m 2016-12-08T01:03:09Z [1] WARNING {agencycomm} Got a redirect 307 from agency endpoint: http+tcp://agency2:8529 location: http://agency:8529/_api/agency/read new forced endpoint: http+tcp://agency:8529/_api/agency/read
←[31mcoordinator_1   |←[0m 2016-12-08T01:03:09Z [1] WARNING {agencycomm} Retrying agency communication at http+tcp://agency:8529 tries: 3
←[33;1mdb2_1           |←[0m 2016-12-08T01:03:10Z [1] INFO {replication} connected to master at tcp://db1:8529, id 180672683468752, version 3.1, last log tick 2116577
←[36;1mdb1_1           |←[0m 2016-12-08T01:03:09Z [1] INFO {replication} connected to master at tcp://db2:8529, id 122177945464438, version 3.1, last log tick 3189617
←[33;1mdb2_1           |←[0m 2016-12-08T01:03:14Z [1] INFO {replication} connected to master at tcp://db1:8529, id 180672683468752, version 3.1, last log tick 2121799
←[36;1mdb1_1           |←[0m 2016-12-08T01:03:23Z [1] ERROR synchronizeOneShard: Could not finalize shard synchronization s100074 { "error" : true, "errorMessage" : "could not contact leader", "response" : { "throw" : throw(msg) {\n    if (this.status >= 400) {\n      throw Object.assign(\n        httperr(this.status,msg || this.message),\n        {details: this}\n      );\n    }\n  } } }←[36;1mdb1_1           |←[0m 2016-12-08T01:03:33Z [1] ERROR cancelReadLockOnLeader: error { "throw" : throw(msg) {\n    if (this.status >= 400) {\n      throw Object.assign(\n        httperr(this.status, msg || this.message),\n        {details: this}\n      );\n    }\n  } }←[36;1mdb1_1           |←[0m 2016-12-08T01:03:33Z [1] ERROR synchronizeOneShard: read lock has timed out for shard s100074
←[36;1mdb1_1           |←[0m 2016-12-08T01:03:33Z [1] ERROR synchronization of local shard 'test/s100074' for central 'test/100072' failed: "Did not work for shard s100074."
←[31mcoordinator_1   |←[0m 2016-12-08T01:03:43Z [1] ERROR Service "/_admin/aardvark" encountered error 500 while handling GET http://localhost:8000/_db/_system/_admin/aardvark/statistics/short?start=1481158981.769815
←[31mcoordinator_1   |←[0m 2016-12-08T01:03:43Z [1] ERROR ArangoError: timeout,cluster node: 'DBServer002', shard: 's100021', endpoint: 'tcp://db2:8529', error: 'timeout in cluster operation' (while executing)
←[31mcoordinator_1   |←[0m 2016-12-08T01:03:43Z [1] ERROR     at Error (native)
←[31mcoordinator_1   |←[0m 2016-12-08T01:03:43Z [1] ERROR     at ArangoStatement.execute (/usr/share/arangodb3/js/server/modules/@arangodb/arango-statement.js:81:16)
←[31mcoordinator_1   |←[0m 2016-12-08T01:03:43Z [1] ERROR     at ArangoDatabase._query (/usr/share/arangodb3/js/server/modules/@arangodb/arango-database.js:90:45)
←[31mcoordinator_1   |←[0m 2016-12-08T01:03:43Z [1] ERROR     at computeStatisticsRaw (/usr/share/arangodb3/js/apps/system/_admin/aardvark/APP/statistics.js:129:21)
←[31mcoordinator_1   |←[0m 2016-12-08T01:03:43Z [1] ERROR     at computeStatisticsShort (/usr/share/arangodb3/js/apps/system/_admin/aardvark/APP/statistics.js:318:3)
←[31mcoordinator_1   |←[0m 2016-12-08T01:03:43Z [1] ERROR     at Route._handler(/usr/share/arangodb3/js/apps/system/_admin/aardvark/APP/statistics.js:520:18)
←[31mcoordinator_1   |←[0m 2016-12-08T01:03:43Z [1] ERROR     at next (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/router/tree.js:385:15)
←[31mcoordinator_1   |←[0m 2016-12-08T01:03:43Z [1] ERROR     at /usr/share/arangodb3/js/node/node_modules/lodash/lodash.js:9378:25
←[31mcoordinator_1   |←[0m 2016-12-08T01:03:43Z [1] ERROR     at Middleware.router.use (/usr/share/arangodb3/js/apps/system/_admin/aardvark/APP/statistics.js:405:3)
←[31mcoordinator_1   |←[0m 2016-12-08T01:03:43Z [1] ERROR     at next (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/router/tree.js:387:15)

I am running docker for windows.

Containers: 8
 Running: 0
 Paused: 0
 Stopped: 8
Images: 2
Server Version: 1.13.0-rc3
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 51371867a01c467f08af739783b8beafc154c4d7
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.8.12-moby
Operating System: Alpine Linux v3.4
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.837 GiB
Name: moby
ID: NI5T:MYUK:MEJG:WCXK:EBFL:2XRV:DTEG:4ZMC:WAXL:OXB6:IUGT:KKXG
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

Any help would be appreciated. 

Thanks,

**Comment by virus2016** [comment]
Ill wait until Frank comes back to me. Thanks guys!

**Comment by fceller** [comment]
@virus2016 please check https://www.arangodb.com/2016/12/starting-arangodb-cluster-easy-way/

**Comment by dothebart** [comment]
The now recommended way to launch clusters is using the ArangoDBStarter - please see https://github.com/arangodb-helper/arangodb - it also supports using this docker container. 
If any more questions arise, feel free to open another ticket or on the starter repository.

**Comment by haseebnaseem** [comment]
@fceller You mentioned kubernetes support in you comment. I know it has been a while so I was wondering If there is some official support for kubernetes right now.

**Comment by dothebart** [comment]
Hi @haseebnaseem - please have a look at this screencast by our community member @sbaugher https://www.arangodb.com/tech-talks/

**Comment by sarendipitee** [comment]
If anyone is looking for a way to run ArangoDB in Swarm here's an example configuration that allows you to scale Agents, DBServers and Controllers independently with `docker service scale` https://github.com/dumstruck/arango-swarm
Hope it's useful for anyone googling this.

**Comment by Simran-B** [comment]
Thanks for sharing @JonDum, will add a link to your repo to our docs: https://github.com/arangodb/docs/pull/542
